File: src/main/java/com/example/chatService/controller/FriendRequestController.java
package com.example.chatService.controller;

import com.example.chatService.dto.FriendRequestSendDto;
import com.example.chatService.security.UserPrincipal;
import com.example.chatService.service.FriendService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/user/friends")
@RequiredArgsConstructor
@Slf4j
public class FriendRequestController {

    private final FriendService friendService;

    @PostMapping("/request")
    public ResponseEntity<?> send(@RequestBody FriendRequestSendDto dto) {
        return ResponseEntity.ok(friendService.sendFriendRequest(dto.getFromUserId(), dto.getToUserId()));
    }

    @GetMapping("/received/{userId}")
    public ResponseEntity<?> getReceived(@PathVariable String userId) {
        return ResponseEntity.ok(friendService.getReceivedRequests(userId));
    }

    @PostMapping("/accept/{id}")
    public ResponseEntity<?> accept(@PathVariable Long id) {
        return ResponseEntity.ok(friendService.acceptFriendRequest(id));
    }

    @PostMapping("/reject/{id}")
    public ResponseEntity<?> reject(@PathVariable Long id) {
        return ResponseEntity.ok(friendService.reject(id));
    }

    @GetMapping("/list/{userId}")
    public ResponseEntity<?> friendList(@PathVariable String userId) {
        return ResponseEntity.ok(friendService.getFriendList(userId));
    }

    @DeleteMapping("/{friendUserId}")
    public ResponseEntity<?> removeFriend(
            @PathVariable String friendUserId,
            Authentication authentication
    ) {
        UserPrincipal principal = (UserPrincipal) authentication.getPrincipal();
        friendService.removeFriend(principal.getId(), friendUserId);

        return ResponseEntity.ok().build();
    }
}

File: src/main/java/com/example/chatService/controller/FriendRequestController.kt
package com.example.chatService.controller

import com.example.chatService.dto.FriendRequestSendDto
import com.example.chatService.security.UserPrincipal
import com.example.chatService.service.FriendService
import org.springframework.http.ResponseEntity
import org.springframework.security.core.Authentication
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/user/friends")
class FriendRequestController(
    private val friendService: FriendService,
) {
    @PostMapping("/request")
    fun send(@RequestBody dto: FriendRequestSendDto): ResponseEntity<Any> =
        ResponseEntity.ok(
            friendService.sendFriendRequest(
                dto.fromUserId ?: throw IllegalArgumentException("fromUserId is required"),
                dto.toUserId ?: throw IllegalArgumentException("toUserId is required"),
            ),
        )

    @GetMapping("/received/{userId}")
    fun getReceived(@PathVariable userId: String): ResponseEntity<Any> =
        ResponseEntity.ok(friendService.getReceivedRequests(userId))

    @PostMapping("/accept/{id}")
    fun accept(@PathVariable id: Long): ResponseEntity<Any> =
        ResponseEntity.ok(friendService.acceptFriendRequest(id))

    @PostMapping("/reject/{id}")
    fun reject(
        @PathVariable id: Long,
        @AuthenticationPrincipal userPrincipal: UserPrincipal,
    ): ResponseEntity<Any> = ResponseEntity.ok(friendService.reject(id, userPrincipal))

    @GetMapping("/list/{userId}")
    fun friendList(@PathVariable userId: String): ResponseEntity<Any> =
        ResponseEntity.ok(friendService.getFriendList(userId))

    @DeleteMapping("/{friendUserId}")
    fun removeFriend(
        @PathVariable friendUserId: String,
        authentication: Authentication,
    ): ResponseEntity<Void> {
        val principal = authentication.principal as UserPrincipal
        friendService.removeFriend(principal.id, friendUserId)
        return ResponseEntity.ok().build()
    }
}
